### ISSUE TITLE
Two --max-length problems with incremental mode

### ISSUE THREAD

**Issue by frank-dittrich** [magnumripper/JohnTheRipper#2785]
This is OK:
```
$ ./john --stdout --incremental --min-length=20 --max-length=20 | head -n 1
MinLen = 20 exceeds MaxLen = 13
```

Without --min-length too large, john should at least warn instead of silently reducing the specified MaxLen:

```
$ ./john --stdout --incremental --max-length=20 | head -n 1
Press 'q' or Ctrl-C to abort, almost any other key for status
123456
$ ./john --stdout --incremental --max-length=20 --verbosity=5| head -n 1
initUnicode(UNICODE, UTF-8/ISO-8859-1)
UTF-8 -> UTF-8 -> UTF-8
Press 'q' or Ctrl-C to abort, almost any other key for status
123456
```

A second problem occurs when --incremental mode is combined with --mask.

```
$ ./john --incremental --stdout --min-length=20 --mask='?w?w' | head -n 1
MinLen = 10 exceeds MaxLen = 6
$ ./john --incremental --stdout --min-length=20 --mask='?wtesttest' | head -n 1
MinLen = 12 exceeds MaxLen = 5
```

This should be allowed.

In the first case, max length should be 26, which would leave 13 for incremental mode.
In the second case, it should be 21 (13 for incremental mode, 8 for "testtest").

BTW: Markov mode seems to work as expected here:
```
$ ./john --stdout --markov --max-length=80 --mask='?w?w'|head -n 1
Warning: Maxlen = 40 is too large (max = 30)
Press 'q' or Ctrl-C to abort, almost any other key for status
MKV start (stats=$JOHN/stats, lvl=200 len=30 pwd=259356431)
chanerererechanererere
```

**Comment by frank-dittrich** [comment]
OTOH:

Instead of exiting with an error code, markov mode just warns when --min-length is too large:
```
n$ ./john --stdout --markov --min-length=70 --mask='?w?w'|head -n 1
Warning: Maxlen = 62 is too large (max = 30)
Warning: minimum length(35) > maximum length(30), minimum length set to 30
Press 'q' or Ctrl-C to abort, almost any other key for status
MKV start (stats=$JOHN/stats, lvl=200 len=30-30 pwd=259356431)
0p 0:00:00:03 0,00% 0p/s
```

I think, incremental mode and Markov mode shouldn't differ here.
In fact, if an invalid --min-length is specified, I'd prefer exiting with an error message.
Should I create a separate Markov mode issue?

**Comment by magnumripper** [comment]
> Should I create a separate Markov mode issue?

I think we can handle it here. It's sort-of the same problem.

**Comment by claudioandre** [comment]
* Unless needed, e.g., `--min-len=1 --max-len=8 -mask=?w?w?d`
  * I failed to find a reason to change the incremental (itself) min_lengh for mask or regex.
  * I failed to find a reason to change the incremental (itself) max_lengh for mask or regex. 
* And mask.c already changes options.req_minlength and options.req_maxlength if needed.

```
diff --git a/src/inc.c b/src/inc.c
index 465cab924..42de4c698 100644
--- a/src/inc.c
+++ b/src/inc.c
@@ -523,32 +523,37 @@ void do_incremental_crack(struct db_main *db, char *mode)
 	max_count = cfg_get_int(SECTION_INC, mode, "CharCount");
 
 	/* Hybrid mask */
-	min_length -= mask_add_len;
-	max_length -= mask_add_len;
 	our_fmt_len -= mask_add_len;
-	if (mask_num_qw > 1) {
-		min_length /= mask_num_qw;
-		max_length /= mask_num_qw;
+	if (mask_num_qw > 1)
 		our_fmt_len /= mask_num_qw;
-	}
 
 #if HAVE_REXGEN
 	/* Hybrid regex */
 	if ((regex = prepare_regex(options.regex, &regex_case, &regex_alpha))) {
-		if (min_length)
-			min_length--;
-		if (max_length)
-			max_length--;
 		if (our_fmt_len)
 			our_fmt_len--;
 	}
 #endif
 
 	/* Command-line can override (narrow) lengths from config file */
-	if (options.req_minlength > min_length)
+	if (options.req_minlength > min_length) {
 		min_length = options.req_minlength;
-	if (options.req_maxlength && options.req_maxlength < max_length)
+
+#if HAVE_REXGEN
+		if (regex)
+			min_length--;
+#endif
+		if (min_length < 0)
+			min_length = 0;
+	}
+
+	if (options.req_maxlength && options.req_maxlength < max_length) {
 		max_length = options.req_maxlength;
+#if HAVE_REXGEN
+		if (regex)
+			max_length--;
+#endif
+	}
 
 	if (min_length > max_length) {
 		log_event("! MinLen = %d exceeds MaxLen = %d",
```

**Comment by magnumripper** [comment]
@claudio I think you are close but not quite there. Here's without your patch:
```
$ ../run/john-b4 -stdout -inc:digits -mask:?w?w?1 -1=ab -max-len=3 
a
b
11a
11b
33a
33b
77a
77b
22a
22b
99a
99b
00a
00b
55a
55b
66a
66b
44a
44b
88a
88b
22p 0:00:00:00 100.00% (2017-09-27 20:30) 1100p/s 88b
```

And here's with it
```
$ ../run/john -stdout -inc:digits -mask:?w?w?1 -1=ab -max-len=3
11a
11b
33a
33b
77a
77b
22a
22b
99a
99b
00a
00b
55a
55b
66a
66b
44a
44b
88a
88b
20p 0:00:00:00 100.00% (2017-09-27 20:31) 2000p/s 88b
```

I believe the former is correct. Right now I can't fathom what's happening so I just put this here for now 😵

**Comment by claudioandre** [comment]
> (inc's first candidate is an empty string, which leads to the first two cands)

So, are you saying we must change min-len when running hybrid mask? So, it will behave the same way as:
```
diff --git a/run/john.conf b/run/john.conf
index 3dea4055f..29273c070 100644
--- a/run/john.conf
+++ b/run/john.conf
@@ -964,7 +964,7 @@ CharCount = 26
 
 [Incremental:Digits]
 File = $JOHN/digits.chr
-MinLen = 1
+MinLen = 0
 MaxLen = 20
 CharCount = 10
```

****

It is fair enough. But, `-inc` itself is not producing the empty string.
```
$ ../run/john -stdout -inc:digits -max-len=1
1
3
7
2
9
0
5
6
4
8
10p 0:00:00:00 100,00% (2017-09-27 20:41) 166.6p/s 8
```

**Comment by magnumripper** [comment]
> ```
> -MinLen = 1
> +MinLen = 0
> ```

I was rather thinking about adding `-min-len=0` which is supposed to do the same as such config change. However, I see now it doesn't (with or without your patch):
```
$ ../run/john -stdout -inc:digits -min-len=0 -max-len=1 
1
3
7
2
9
0
5
6
4
8
10p 0:00:00:00 100.00% (2017-09-28 07:49) 1000p/s 8
```
This is a(nother) bug.

**Comment by claudioandre** [comment]
> I was rather thinking about adding --min-len=0

It is not working because of:
```
	if (options.req_minlength > min_length)
		min_length = options.req_minlength;
```

See 3e58a0aade8dfddff069f85be281eaad16229b87 and the "narrow" comment.

**Comment by magnumripper** [comment]
Ah, OK. Well for max-len we have to use such limitation, but for min-len we should be able to over-ride both ways.

**Comment by claudioandre** [comment]
Would be interesting to warn if `--max-len=n` is valid". Since `options.req_maxlength` can be "adjusted" by mask.c, we need to know whether req_maxlength was informed by the user or computed by mask.c:
```
diff --git a/src/inc.c b/src/inc.c
index 72e3dfb6b..1fa937192 100644
--- a/src/inc.c
+++ b/src/inc.c
@@ -547,12 +547,22 @@ void do_incremental_crack(struct db_main *db, char *mode)
                        min_length = 0;
        }
 
-       if (options.req_maxlength && options.req_maxlength < max_length) {
-               max_length = options.req_maxlength;
+       if (options.req_maxlength && maxlength_informed) {
+               if (options.req_maxlength < max_length) {
+                       max_length = options.req_maxlength;
 #if HAVE_REXGEN
-               if (regex)
-                       max_length--;
+                       if (regex)
+                               max_length--;
 #endif
+
+               } else {
+                       log_event("! MaxLen = %d exceeds the config file limit of %d",
+                           options.req_maxlength, max_length);
+                       if (john_main_process)
+                               fprintf(stderr, "MaxLen = %d exceeds the config file "
+                                   "limit of %d\n", options.req_maxlength, max_length);
+                       error();
+               }
        }
 
        if (min_length > max_length) {
```

**Comment by magnumripper** [comment]
> Since options.req_maxlength can be "adjusted" by mask.c

Yes, and actually we should allow -max-len command line bumping Incremental's maxlen beyond the conf too.

**Comment by magnumripper** [comment]
...so we don't need to know who set it, just add such warning!

**Comment by claudioandre** [comment]
I prefer something like this:
* valid: `../run/john --incremental --stdout --min-length=40 --mask='?w?w?w' | head -n 5`
  * MaxLen = 13
* complains and exit (as upstream does): `../run/john --stdout --incremental --max-length=20`
  * MaxLen = 20 exceeds the config file limit of 13
```
diff --git a/src/inc.c b/src/inc.c
index 72e3dfb6b..a32b24849 100644
--- a/src/inc.c
+++ b/src/inc.c
@@ -547,12 +547,22 @@ void do_incremental_crack(struct db_main *db, char *mode)
 			min_length = 0;
 	}
 
-	if (options.req_maxlength && options.req_maxlength < max_length) {
-		max_length = options.req_maxlength;
+	if (options.req_maxlength && !maxlength_computed) {
+		if (options.req_maxlength < max_length) {
+			max_length = options.req_maxlength;
 #if HAVE_REXGEN
-		if (regex)
-			max_length--;
+			if (regex)
+				max_length--;
 #endif
+
+		} else {
+			log_event("! MaxLen = %d exceeds the config file limit of %d",
+			    options.req_maxlength, max_length);
+			if (john_main_process)
+				fprintf(stderr, "MaxLen = %d exceeds the config file "
+				    "limit of %d\n", options.req_maxlength, max_length);
+			error();
+		}
 	}
 
 	if (min_length > max_length) {
diff --git a/src/mask.c b/src/mask.c
index 51b015a85..e1f9312e0 100644
--- a/src/mask.c
+++ b/src/mask.c
@@ -48,7 +48,7 @@ static unsigned long long rec_cl, cand_length;
 static struct fmt_main *mask_fmt;
 static int mask_bench_index;
 static int parent_fix_state_pending;
-int mask_add_len, mask_num_qw, mask_cur_len;
+int mask_add_len, mask_num_qw, mask_cur_len, maxlength_computed = 0;
 
 /*
  * This keeps track of whether we have any 8-bit in our non-hybrid mask.
@@ -2008,6 +2008,9 @@ void mask_init(struct db_main *db, char *unprocessed_mask)
 
 	fmt_maxlen = mask_fmt->params.plaintext_length;
 
+	// Track to know whether max-length is informed or computed
+	maxlength_computed = (options.req_maxlength == 0);
+
 	if (options.req_minlength >= 0 && !options.req_maxlength)
 		options.req_maxlength = fmt_maxlen;
 	else if (options.req_maxlength > 0 && options.req_minlength == -1)
diff --git a/src/mask.h b/src/mask.h
index 7223ad629..8ed4ed6cc 100644
--- a/src/mask.h
+++ b/src/mask.h
@@ -118,4 +118,7 @@ extern unsigned long long mask_parent_keys;
 /* Current length when pure mask mode iterates over lengths */
 extern int mask_cur_len;
 
+/* Set if max-length is computed (as opposed to informed by an user) */
+extern int maxlength_computed;
+
 #endif
```

**Comment by frank-dittrich** [comment]
s/informed by an user/specified by the user/

**Comment by frank-dittrich** [comment]
This recent change to incremental mode improved the old behaviour.

```
 $ ./john --stdout --incremental --min-length=20 --max-length=20 | head -n 1
MinLen = 20 exceeds MaxLen = 13
```
into
```
$ ./john --stdout --incremental --min-length=20 --max-length=20 | head -n 1
Press 'q' or Ctrl-C to abort, almost any other key for status
aaaaaaaaaaaaaaaaaaaa
```

The new limit seems to be 24, which is the compile-time limit:
```
$ ./john --stdout --incremental --min-length=20 --max-length=50 | head -n 1
MaxLen = 50 exceeds the compile-time limit of 24
```

Now the MinLen checks:
```
$ ./john --stdout --incremental --min-length=50 | head -n 1
MinLen = 50 exceeds MaxLen = 13
$ ./john --stdout --incremental --min-length=20 | head -n 1
MinLen = 20 exceeds MaxLen = 13
```
If the MaxLen from the config file is smaller than the compile time limit, it would be user friendly to point out that --max-length can be increased up to the compile time limit.

The following problem seems to be related:
```
$ ./john --incremental --stdout --min-length=48 --mask='?w?w' | head -n 1
MinLen = 24 exceeds MaxLen = 13
```
Because with --max-length, it works:
```
$ ./john --incremental --stdout --min-length=48 --max-length=48 --mask='?w?w' | head -n 1
Press 'q' or Ctrl-C to abort, almost any other key for status
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
```

**Comment by frank-dittrich** [comment]
A minor problem remains for --incremental and a mask with more than one `?w` or `?W`:
The way mask mode computes the incremental mode min length may produce candidates which are shorter than the requested min. length:
```
$ ./john --incremental --stdout --min-length=3 --max-length=4 --mask='?w?w' |head -n 4
Press 'q' or Ctrl-C to abort, almost any other key for status
meme
mjmj
11
hihi
$ ./john --incremental --stdout --min-length=1 --max-length=4 --mask='?w?w' |head -n 4
Press 'q' or Ctrl-C to abort, almost any other key for status
meme
mjmj

11
```

**Comment by frank-dittrich** [comment]
No, my test was flawed: Hybrid mode min. length calculation for --markov only avoids min. length 1, but not 3:
```
$ ./john --markov=400 --stdout --min-length=3 --max-length=4 --mask='?w?w' | wc -l
Press 'q' or Ctrl-C to abort, almost any other key for status
MKV start (stats=$JOHN/stats, lvl=400 len=1-2 pwd=23219)
23230p 0:00:00:00 100,00% (2017-11-12 14:27) 774333p/s ��
23412
$ ./john --markov=400 --stdout --min-length=4 --max-length=4 --mask='?w?w' | wc -l
Press 'q' or Ctrl-C to abort, almost any other key for status
MKV start (stats=$JOHN/stats, lvl=400 len=2-2 pwd=23219)
23053p 0:00:00:00 99,24% (ETA: 14:27:50) 1152Kp/s ^n^n
23233
```

**Comment by magnumripper** [comment]
Several fixes has been made (some specifically addressing this issue without referring to it, I believe) and this issue is starting to get hard to follow. @frank-dittrich please open a new issue if you still see something amiss.
